sklearn_numba_dpex/kmeans/drivers.py

```diff
@@ -0,0 +1,459 @@
+import math
+import warnings
+
+import numpy as np
+import dpctl
+from sklearn.exceptions import DataConversionWarning
+
+from sklearn_numba_dpex.utils._device import _DeviceParams
+
+from sklearn_numba_dpex.kmeans.kernels import (
+    make_initialize_to_zeros_1dim_kernel,
+    make_initialize_to_zeros_2dim_kernel,
+    make_initialize_to_zeros_3dim_kernel,
+    make_copyto_kernel,
+    make_broadcast_division_kernel,
+    make_center_shift_kernel,
+    make_half_l2_norm_dim0_kernel,
+    make_sum_reduction_1dim_kernel,
+    make_sum_reduction_2dim_kernel,
+    make_sum_reduction_3dim_kernel,
+    make_fused_fixed_window_kernel,
+    make_assignment_fixed_window_kernel,
+)
+
+
+def _check_power_of_2(e):
+    if e != 2 ** (math.log2(e)):
+        raise ValueError(f"Expected a power of 2, got {e}")
+    return e
+
+
+class LLoydKMeansDriver:
+    """GPU optimized implementation of Lloyd's k-means.
+
+    Inspired by the "Fused Fixed" strategy exposed in:
+    Kruliš, M., & Kratochvíl, M. (2020, August). Detailed analysis and
+    optimization of CUDA K-means algorithm. In 49th International
+    Conference on Parallel Processing-ICPP (pp. 1-11).
```

Comment:
Small nit: Docstrings can potentially follow [numpydoc style](https://numpydoc.readthedocs.io/en/latest/) used by scikit-learn.

In that case, this reference can be moved in a [`References` section](https://numpydoc.readthedocs.io/en/latest/format.html#references) and be cited here.

Reply:
It's unfortunate that this paper is not open access and we cannot include direct URL to a publicly downloadable PDF of the paper.

However we should definitely link to:

- the github repo of the authors of the paper with the original CUDA implementation:
  https://github.com/krulis-martin/cuda-kmeans

- a slide deck with schemas to intuitively explain the main computation:
  https://jnamaral.github.io/icpp20/slides/Krulis_Detailed.pdf